### PR TITLE
Correct DQL `INSTANCE OF` to filter all possible child classes

### LINKED ISSUE #4646: DDC-3800: [GH-1441] [QUERY] "INSTANCE OF" now behaves correctly with subclasses
Jira issue originally created by user @doctrinebot:

This issue is created automatically through a Github pull request on behalf of taueres:

Url: https://github.com/doctrine/doctrine2/pull/1441

Message:

There was a bug in the "INSTANCE OF" operator as described in
https://groups.google.com/forum/#!topic/doctrine-user/B8raq8CNMgg

"INSTANCE OF" was not taking into account subclasses.
It was merely translating the class to its discriminator.
This is not correct since the class can have subtypes and those
are, indeed, still instance of the superclass.
Also, classes may not have a discriminator (e.g. abstract classes).

This commit also provide useful tests to avoid regression.

### PR DESCRIPTION
**Pull request by Jean85**
This is the rebased version of #1441, which fixes #4646.
Here is the original text:

> There was a bug in the "INSTANCE OF" operator as described in
> https://groups.google.com/forum/#!topic/doctrine-user/B8raq8CNMgg
> 
> "INSTANCE OF" was not taking into account subclasses.
> It was merely translating the class to its discriminator.
> This is not correct since the class can have subtypes and those
> are, indeed, still instance of the superclass.
> Also, classes may not have a discriminator (e.g. abstract classes).
> 
> This commit also provides useful tests to avoid regression.

### CODE REVIEW

File: tests/Doctrine/Tests/ORM/Functional/InstanceOfTest.php [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    use Doctrine\Tests\OrmFunctionalTestCase;
+
+    class InstanceOfTest extends OrmFunctionalTestCase
+    {
+        protected function setUp()
+        {
+            parent::setUp();
+
+            $this->_schemaTool->createSchema(array(
```
Comment (phansys) [inline tests/Doctrine/Tests/ORM/Functional/InstanceOfTest.php]: Please, use short array syntax instead.

File: tests/Doctrine/Tests/ORM/Functional/InstanceOfAbstractTest.php [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+            $dql = 'SELECT p FROM Doctrine\Tests\ORM\Functional\InstanceOfAbstractTest\Person p
+                    WHERE p INSTANCE OF Doctrine\Tests\ORM\Functional\InstanceOfAbstractTest\Person';
+            $query = $this->_em->createQuery($dql);
+            $result = $query->getResult();
+
+            $this->assertCount(1, $result);
+
+            foreach ($result as $r) {
+                $this->assertInstanceOf('Doctrine\Tests\ORM\Functional\InstanceOfAbstractTest\Person', $r);
```
Comment (phansys) [inline tests/Doctrine/Tests/ORM/Functional/InstanceOfAbstractTest.php]: Please, use `::class` constant instead of hardcoding the QCN here.

File: tests/Doctrine/Tests/ORM/Functional/InstanceOfAbstractTest.php [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                $this->assertInstanceOf('Doctrine\Tests\ORM\Functional\InstanceOfAbstractTest\Employee', $r);
+                $this->assertEquals('bar', $r->getName());
```
Comment (phansys) [inline tests/Doctrine/Tests/ORM/Functional/InstanceOfAbstractTest.php]: Please, use `assertSame()` instead.

**Review by phansys — CHANGES_REQUESTED** [review]
Status: Needs work.

**Comment by Jean85** [comment]
All done, thanks for the review @phansys

**Review by lcobucci — CHANGES_REQUESTED** [review]
Can you please extract some methods from `SqlWalker#walkInstanceOfExpression()` it was already messy and now it's a bit harder to understand things properly.

Also could you try to relate the tests, and entities that each test use, to the ticket (as we have on `Functional\Ticket` classes) just to keep the "standard" we're having...

**Comment by Jean85** [comment]
Thanks for the review @lcobucci!
I hope that I addressed al your concerns with my latest commits.

**Comment by Jean85** [comment]
A lot of tests are failing, probably for an unrelated issue... should I rebase? Or merge master?

File: lib/Doctrine/ORM/Query/SqlWalker.php [hunk 4]
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
...
+            // Include discriminators for parameter class and its subclass
+            $metadata = $this->em->getClassMetadata($parameter);
+            $hierarchyClasses = $metadata->subClasses;
+            $hierarchyClasses[] = $metadata->name;
+
+            foreach ($hierarchyClasses as $class) {
+                $currentMetadata = $this->em->getClassMetadata($class);
+                $currentDiscriminator = $currentMetadata->discriminatorValue;
+
+                if (is_string($currentDiscriminator) && ! array_key_exists($currentDiscriminator, $discriminators)) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: A discriminator may be also an integer or similar - why do you want this to only work with `string`?

File: lib/Doctrine/ORM/Query/SqlWalker.php [hunk 5]
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
...
+            if ($parameter instanceof AST\InputParameter) {
+                $this->rsm->addMetadataParameterMapping($parameter->name, 'discriminatorValue');
+
+                $sqlParameterList[] = $this->walkInputParameter($parameter);
+
+                continue;
+            }
+
+            // Trim first backslash
+            $parameter = ltrim($parameter, '\\');
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: You should do `$this->em->getClassMetadata($parameter);` here - this already provides the `ltrim()` logic for you (internally)

File: lib/Doctrine/ORM/Query/SqlWalker.php [hunk 6]
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
...
+
+            if ($parameter !== $discrClass->name && ! array_key_exists($parameter, $knownSubclasses)) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: `$metadata->getClass()->isSubclassOf(...)` maybe?

Comment (Ocramius) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: This also has the effect of getting rid of `INSTANCE OF` checks where `INSTANCE OF` is pointing at the topmost class (which is most likely a bug in userland code)

File: lib/Doctrine/ORM/Query/SqlWalker.php [hunk 7]
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
...
+            $metadata = $this->em->getClassMetadata($parameter);
+            $hierarchyClasses = $metadata->subClasses;
+            $hierarchyClasses[] = $metadata->name;
+
+            foreach ($hierarchyClasses as $class) {
+                $currentMetadata = $this->em->getClassMetadata($class);
+                $currentDiscriminator = $currentMetadata->discriminatorValue;
+
+                if (is_string($currentDiscriminator) && ! array_key_exists($currentDiscriminator, $discriminators)) {
+                    $discriminators[$currentDiscriminator] = true;
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: Consider using `null` if that `true` has no meaning

File: lib/Doctrine/ORM/Query/SqlWalker.php [hunk 8]
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
...
+                }
+            }
+        }
+
+        foreach (array_keys($discriminators) as $dis) {
+            $sqlParameterList[] = $this->conn->quote($dis);
+        }
+
+        return '(' . implode(', ', $sqlParameterList) . ')';
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: According to this logic, the list may be empty, leading to a crash

File: lib/Doctrine/ORM/Query/SqlWalker.php [hunk 9]
```diff
@@ -2300,4 +2271,53 @@ public function walkResultVariable($resultVariable)
...
+                $currentDiscriminator = $currentMetadata->discriminatorValue;
+
+                if (is_string($currentDiscriminator) && ! array_key_exists($currentDiscriminator, $discriminators)) {
+                    $discriminators[$currentDiscriminator] = true;
+                }
+            }
+        }
+
+        foreach (array_keys($discriminators) as $dis) {
+            $sqlParameterList[] = $this->conn->quote($dis);
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: `array_map([$this->conn, 'quote'], array_keys($discriminators))`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php [hunk 10]
```diff
@@ -0,0 +1,114 @@
+<?php
+
+namespace Doctrine\Tests\ORM\Functional\Ticket {
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php]: Please put the entire file under a single namespace

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+     * @Table(name="instance_of_multi_level_engineer")
+     */
+    class Engineer extends Employee
+    {
+        /**
+         * @Column(type="string")
+         */
+        private $specialization;
+
+        public function getSpecialization()
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php]: Remove all accessors that aren't strictly needed: prefer public properties if they are needed (although note that these tests don't really need the properties to work, so you should drop everything but the identifier)

File: tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php [hunk 12]
```diff
@@ -502,7 +502,7 @@ public function testSupportsInstanceOfExpressionsInWherePart()
     {
         $this->assertSqlGeneration(
             "SELECT u FROM Doctrine\Tests\Models\Company\CompanyPerson u WHERE u INSTANCE OF Doctrine\Tests\Models\Company\CompanyEmployee",
-            "SELECT c0_.id AS id_0, c0_.name AS name_1, c0_.discr AS discr_2 FROM company_persons c0_ WHERE c0_.discr IN ('employee')"
+            "SELECT c0_.id AS id_0, c0_.name AS name_1, c0_.discr AS discr_2 FROM company_persons c0_ WHERE c0_.discr IN ('manager', 'employee')"
```
Comment (Ocramius) [inline tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php]: Ouch, this is a bad bug. Good catch!

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: lib/Doctrine/ORM/Query/SqlWalker.php (hunk 5 above)
Comment (Jean85) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: Right! I was also already doing the `getClassMetadata` a few lines after, I'll just move that up!

File: lib/Doctrine/ORM/Query/SqlWalker.php (hunk 4 above)
Comment (Jean85) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: I didn't know! I'll do a `!== null` then.

File: tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php (hunk 12 above)
Comment (Jean85) [inline tests/Doctrine/Tests/ORM/Query/SelectSqlGenerationTest.php]: All credits to @taueres !

File: lib/Doctrine/ORM/Query/SqlWalker.php (hunk 9 above)
Comment (Jean85) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: We may already have values inside `$sqlParameterList`, see [line 2289](https://github.com/doctrine/doctrine2/pull/6392/files#diff-0c8825d1265a66bb8b20f3d99c276961R2289), so I think it's better to do it like that.

File: lib/Doctrine/ORM/Query/SqlWalker.php (hunk 8 above)
Comment (Jean85) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: How that could happen? Since [I check if the class is unrelated at line 2297](https://github.com/doctrine/doctrine2/pull/6392/files#diff-0c8825d1265a66bb8b20f3d99c276961R2297), at least one discriminator should pop up, or an exception be thrown...

**Comment by Jean85** [comment]
Ok seems that I've addressed all your concerns, @Ocramius @lcobucci 
Build is green with low deps, other failures seems unrelated, maybe they come from master?

**Comment by taueres** [comment]
Hi! Thank you very much for resuming this PR, I really appreciate that!
I checked this fix once more but unfortunately the issue still occurs when using `INSTANCE OF` with parameters.

Here is an example:
```php
$dql = 'SELECT p FROM Person p WHERE p INSTANCE OF :parameter';
$query = $em->createQuery($dql);
$query->setParameter('parameter', $em->getClassMetadata(Person::class));

// $result contains only direct instances of `Person`, no subclasses
$result = $query->getResult();
```

**Comment by taueres** [comment]
I wrote a failing test case for the issue I mentioned before: https://gist.github.com/taueres/05ff035bb2f72cddd7711b7f93cd9d6a

File: lib/Doctrine/ORM/Query/SqlWalker.php (hunk 9 above)
Comment (Ocramius) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: Then just `array_merge()` with it

File: lib/Doctrine/ORM/Query/SqlWalker.php (hunk 8 above)
Comment (Ocramius) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: Yeah, but that flow makes it hard to follow: ideally we should `array_map()` over `$instanceOfExpr->value`, but let's see where the code goes after the additional tests provided by @taueres

**Comment by Jean85** [comment]
Thanks @taueres!
I've found that your test case fails also when doing this:

```php
$dql = 'SELECT p FROM Doctrine\Tests\ORM\Functional\Ticket\PersonTicket4646Parametric p
        WHERE p INSTANCE OF :parameter';
$query = $this->_em->createQuery($dql);
$query->setParameter(
    'parameter',
    PersonTicket4646Parametric::class
);
```
Do we want to check for this additional uses, @Ocramius ?

**Comment by Ocramius** [comment]
```php
$query->setParameter(
    'parameter',
    PersonTicket4646Parametric::class
);
```

This currently requires binding a `ClassMetadata` instance:

```php
$query->setParameter(
    'parameter',
    $em->getClassMetadata(PersonTicket4646Parametric::class)
);
```

We can fix it, yes

**Comment by Jean85** [comment]
Ok I've added the failing test case. I'll try to work on that soon; @taueres if you have any tips, I'm all ears! :smile:

**Comment by taueres** [comment]
Unfortunately solving the parameter issue is not as easy as static DQL queries.
The main problem is that **Doctrine is always translating each DQL parameter to one single SQL parameter**.
For `INSTANCE OF`, this 1-to-1 relationship is not always true: one class might be mapped to multiple discriminators.

The issue is not easy to address because the parameter values could still be undefined while Doctrine is parsing the query.

Think about the following scenario:
```php
$dql = 'SELECT p FROM Person p WHERE p INSTANCE OF :parameter';
$sql = $query->getSQL();
// $sql is something like "SELECT * FROM person WHERE person.kind IN (?)"
// which is WRONG because we don't know yet how many SQL parameters are needed.
```

Do you have any idea of how we could tackle this problem?
Is DBAL supporting `IN` operator with array parameters?

**Comment by Ocramius** [comment]
> Is DBAL supporting `IN operator with array parameters?

Yep, but it only supports [these two types of bindings](https://github.com/doctrine/dbal/blob/03e17189c48a92bff9c5fd7d9242d13a15780d15/lib/Doctrine/DBAL/Connection.php#L73-L85), so custom discriminators may be a problem

**Comment by taueres** [comment]
Okay, I have a solution! `INSTANCE OF` with parameters now works correctly on my branch https://github.com/taueres/doctrine2/tree/fix-instance-of-subclasses

It still needs some minor improvements, e.g. duplicated methods.

How should I proceed with this PR? Should I create a new one from my branch?

**Comment by Ocramius** [comment]
@taueres I suggest waiting for @Jean85 to merge the changes into his branch

**Comment by Jean85** [comment]
I've merged your fixes @taueres, thanks a lot!
Now the only issue that I see is the duplication of the `getAllDiscriminators(ClassMetadata $classMetadata)` method in both the `Query` and `SqlWalker` classes. Any suggestion on how to address that? It uses `$this->getEntityManager()->getClassMetadata($class)`, which I think is the biggest hurdle to overcome, otherwise it could be moved inside the `ClassMetadata` class.

**Comment by taueres** [comment]
To solve the duplication issue we could create something like:
```php
class HierarchyDiscriminatorResolver
{
    public static function resolveDiscriminatorsForClass(ClassMetadata $rootClassMetadata, EntityManagerInterface $entityManager)
    {
        /* Duplicate code goes here */
        /* return list of discriminator */
    }
}
```

In the proposed solution, the `EntityManager` dependency is needed in order to retrieve `ClassMetadata` for subClasses.

I would not put such logic into `ClassMetadata` because atm `ClassMetadata` is not coupled to `EntityManager`.

**Comment by Jean85** [comment]
This seems a feasible solution for me... @Ocramius ? What do you think?

**Comment by Ocramius** [comment]
@Jean85 as long as the class is marked as `@internal`, sure :+1: 

Totally missed these comments before btw.

**Comment by Jean85** [comment]
Don't worry @Ocramius, I was on vacation anyway :smile: 
Deduplication of code done, build is green, this is ready for a final review!

File: lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+use Doctrine\Common\Persistence\Mapping\ClassMetadata;
+use Doctrine\ORM\EntityManagerInterface;
+
+/**
+ * Class HierarchyDiscriminatorResolver
+ * @package Doctrine\ORM\Utility
+ * @internal This class exists only to avoid code duplication, do not reuse it externally
+ */
+class HierarchyDiscriminatorResolver
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php]: This class needs to be tested in isolation

File: lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php [hunk 14]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Doctrine\ORM\Utility;
+
+use Doctrine\Common\Persistence\Mapping\ClassMetadata;
+use Doctrine\ORM\EntityManagerInterface;
+
+/**
+ * Class HierarchyDiscriminatorResolver
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php]: Remove this line

File: lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+ * @package Doctrine\ORM\Utility
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php]: Remove this line

File: lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php (hunk 13 above)
Comment (Ocramius) [inline lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php]: `final`

File: lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+ * @internal This class exists only to avoid code duplication, do not reuse it externally
+ */
+class HierarchyDiscriminatorResolver
+{
+    public static function resolveDiscriminatorsForClass(
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php]: Add a `private` constructor

File: lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+        ClassMetadata $rootClassMetadata,
+        EntityManagerInterface $entityManager
+    ) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php]: Return type

File: lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+        $hierarchyClasses = $rootClassMetadata->subClasses;
+        $hierarchyClasses[] = $rootClassMetadata->name;
+
+        $discriminators = [];
+        foreach ($hierarchyClasses as $class) {
```
Comment (Ocramius) [inline lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php]: Empty line before logical blocks

File: lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php (hunk 16 above)
Comment (Ocramius) [inline lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php]: An explanation of the method is needed

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: lib/Doctrine/ORM/Query/SqlWalker.php [hunk 19]
```diff
@@ -2297,4 +2269,37 @@ public function walkResultVariable($resultVariable)
...
         return $resultAlias;
     }
+
+    /**
+     * @param ClassMetadataInfo $rootClass
+     * @param AST\InstanceOfExpression $instanceOfExpr
+     * @return string The list in parentheses of valid child discriminators from the given class
+     * @throws QueryException
+     */
+    private function getChildDiscriminatorsFromClassMetadata(ClassMetadataInfo $rootClass, AST\InstanceOfExpression $instanceOfExpr)
```
Comment (lcobucci) [inline lib/Doctrine/ORM/Query/SqlWalker.php]: Please add the  return type declaration

**Comment by taueres** [comment]
According to the DQL spec. Users can provide multiple classes for an INSTANCE OF statement.
```
InstanceOfExpression     ::= IdentificationVariable ["NOT"] "INSTANCE" ["OF"] (InstanceOfParameter | "(" InstanceOfParameter {"," InstanceOfParameter}* ")")
```
Do we have some tests covering this specific use case?

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php [hunk 20]
```diff
@@ -0,0 +1,64 @@
+<?php
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+class Ticket4646InstanceOfAbstractTest extends OrmFunctionalTestCase
+{
+    protected function setUp()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php]: `: void`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        parent::setUp();
+
+        $this->_schemaTool->createSchema([
+            $this->_em->getClassMetadata(PersonTicket4646Abstract::class),
+            $this->_em->getClassMetadata(EmployeeTicket4646Abstract::class),
+        ]);
+    }
+
+    public function testInstanceOf()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php]: `: void`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        $this->_em->persist(new EmployeeTicket4646Abstract());
+        $this->_em->flush();
+
+        $dql = 'SELECT p FROM Doctrine\Tests\ORM\Functional\Ticket\PersonTicket4646Abstract p
+                WHERE p INSTANCE OF Doctrine\Tests\ORM\Functional\Ticket\PersonTicket4646Abstract';
+        $query = $this->_em->createQuery($dql);
+        $result = $query->getResult();
+
+        $this->assertCount(1, $result);
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfAbstractTest.php]: Please use `self::assert*` for new tests

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php [hunk 23]
```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+class Ticket4646InstanceOfMultiLevelTest extends OrmFunctionalTestCase
+{
+    protected function setUp()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php]: `: void`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+        parent::setUp();
+
+        $this->_schemaTool->createSchema([
+            $this->_em->getClassMetadata(PersonTicket4646MultiLevel::class),
+            $this->_em->getClassMetadata(EmployeeTicket4646MultiLevel::class),
+            $this->_em->getClassMetadata(EngineerTicket4646MultiLevel::class),
+        ]);
+    }
+
+    public function testInstanceOf()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php]: `: void`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+        $this->_em->persist(new EmployeeTicket4646MultiLevel());
+        $this->_em->persist(new EngineerTicket4646MultiLevel());
+        $this->_em->flush();
+
+        $dql = 'SELECT p FROM Doctrine\Tests\ORM\Functional\Ticket\PersonTicket4646MultiLevel p
+                WHERE p INSTANCE OF Doctrine\Tests\ORM\Functional\Ticket\PersonTicket4646MultiLevel';
+        $query = $this->_em->createQuery($dql);
+        $result = $query->getResult();
+
+        $this->assertCount(3, $result);
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php]: Here too (`self::assert*`)

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php [hunk 26]
```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+class Ticket4646InstanceOfParametricTest extends OrmFunctionalTestCase
+{
+    protected function setUp()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php]: `: void`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+    {
+        parent::setUp();
+        $this->_schemaTool->createSchema([
+            $this->_em->getClassMetadata(PersonTicket4646Parametric::class),
+            $this->_em->getClassMetadata(EmployeeTicket4646Parametric::class),
+        ]);
+    }
+
+    public function testInstanceOf()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php]: `: void`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+        $this->_em->flush();
+        $dql = 'SELECT p FROM Doctrine\Tests\ORM\Functional\Ticket\PersonTicket4646Parametric p
+                WHERE p INSTANCE OF :parameter';
+        $query = $this->_em->createQuery($dql);
+        $query->setParameter(
+            'parameter',
+            $this->_em->getClassMetadata(PersonTicket4646Parametric::class)
+        );
+        $result = $query->getResult();
+        $this->assertCount(2, $result);
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php]: `self::assert*` here too

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php [hunk 29]
```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace Doctrine\Tests\ORM\Functional\Ticket;
+
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+class Ticket4646InstanceOfTest extends OrmFunctionalTestCase
+{
+    protected function setUp()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php]: `: void`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php [hunk 30]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        parent::setUp();
+
+        $this->_schemaTool->createSchema([
+            $this->_em->getClassMetadata(PersonTicket4646::class),
+            $this->_em->getClassMetadata(EmployeeTicket4646::class),
+        ]);
+    }
+
+    public function testInstanceOf()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php]: `: void`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php [hunk 31]
```diff
@@ -0,0 +1,77 @@
...
+class PersonTicket4646MultiLevel
+{
+    /**
+     * @Id()
+     * @GeneratedValue()
+     * @Column(type="integer")
+     */
+    private $id;
+
+    public function getId()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfMultiLevelTest.php]: `: ?int`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php [hunk 32]
```diff
@@ -0,0 +1,67 @@
...
+class PersonTicket4646Parametric
+{
+    /**
+     * @Id()
+     * @GeneratedValue()
+     * @Column(type="integer")
+     */
+    private $id;
+
+    public function getId()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfParametricTest.php]: `: ?int`

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php [hunk 33]
```diff
@@ -0,0 +1,66 @@
...
+        $this->_em->persist(new PersonTicket4646());
+        $this->_em->persist(new EmployeeTicket4646());
+        $this->_em->flush();
+
+        $dql = 'SELECT p FROM Doctrine\Tests\ORM\Functional\Ticket\PersonTicket4646 p
+                WHERE p INSTANCE OF Doctrine\Tests\ORM\Functional\Ticket\PersonTicket4646';
+        $query = $this->_em->createQuery($dql);
+        $result = $query->getResult();
+
+        $this->assertCount(2, $result);
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php]: Here too

File: tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php [hunk 34]
```diff
@@ -0,0 +1,66 @@
...
+class PersonTicket4646
+{
+    /**
+     * @Id()
+     * @GeneratedValue()
+     * @Column(type="integer")
+     */
+    private $id;
+
+    public function getId()
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/Ticket4646InstanceOfTest.php]: `: ?int`

**Review by lcobucci — DISMISSED** [review]
Some nitpicking 😄

File: lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php (hunk 17 above)
Comment (Jean85) [inline lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php]: Sorry, I forgot that the target branch is 7+! Good!

File: lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php (hunk 18 above)
Comment (Jean85) [inline lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php]: You should enforce this in the Lint stage of the CI!

Comment (lcobucci) [inline lib/Doctrine/ORM/Utility/HierarchyDiscriminatorResolver.php]: We're going to :wink

**Comment by Jean85** [comment]
@lcobucci @Ocramius done as per review!

**Comment by Jean85** [comment]
I've also added a new functional test to cover the case of multiple parameters passed to the `INSTANCEOF`, as suggested by @taueres

**Review by Ocramius — APPROVED** [review]

**Review by Ocramius — APPROVED** [review]
LGTM :+1: :ship:

**Comment by Ocramius** [comment]
Thanks @Jean85 @taueres and @phansys!
